gcp_variant_transforms/vcf_to_bq.py

```diff
@@ -233,27 +234,32 @@ def run(argv=None):
   processed_variant_factory = processed_variant.ProcessedVariantFactory(
       header_fields,
       known_args.split_alternate_allele_info_fields,
-      known_args.annotation_fields)
+      known_args.annotation_fields,
+      metrics_util.CounterFactory())
```

Comment:
CounterFactory is completely stateless and can be instantiated many times.
The reason it is passed is exactly to avoid direct instantiation inside the `ProcessedVariantFactory` class to avoid tying it to a particular implementation. Instead, `ProcessedVariantFactory` only knows/expects the interface of the factory method with no information of what implementation is actually used.

I think this is the root of our difference on this design. Direct instantiation of classes in a constructor, specially those that we like to hide their implementation (e.g., Counters and its Factory in this case) is not a good idea. I shared [this writing](http://misko.hevery.com/code-reviewers-guide/flaw-constructor-does-real-work/) before in a review thread as well and I appreciate if you take a look (at least glance through it and check the last section before codes/comments which is "Fixing the flaw"). As I said before, I know that not all of it applies to Python but in general, I have found *injecting factories* a good design pattern to have modular code.

Now, you may disagree to the general pattern or the particular usage here. In either case, please let me know and I will reduce the whole beam_util module to a single single `CounterWrapper` class, removing all other factory/interface/no no-op details (if I want to instantiate a particular implementation why even bother with factories?). I don't think that is a good design but let's move on. I don't believe there is a "best design" and I don't want this PR to be blocked more. I am fine either way.

Reply:
Thanks for sharing the doc. Sorry I missed it originally. I like the factory design pattern and now see your point about keeping the constructor light. I'd still prefer to create one instance of the factory and passing that in to any module that wants to use counters (instead of each module creating their own instance in the method call), but feel free to leave as is for now.

Reply:
Done.